File: apps/go-sdk/tools.go
package firecrawl

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// ExchangeExecutionError preserves the request ID for retrying the same payload.
type ExchangeExecutionError struct {
	RequestID string
	Err       error
}

func (e *ExchangeExecutionError) Error() string {
	return fmt.Sprintf("%v (request ID: %s)", e.Err, e.RequestID)
}
func (e *ExchangeExecutionError) Unwrap() error { return e.Err }

func (c *Client) ScrapeExchange(ctx context.Context, calls []ExchangeCall, opts *ExchangeOptions) (*ExchangeScrapeData, error) {
	if len(calls) == 0 {
		return nil, &FirecrawlError{Message: "at least one exchange call is required"}
	}
	if len(calls) > 10 {
		return nil, &FirecrawlError{Message: "at most 10 exchange calls are allowed per request"}
	}
	for i, call := range calls {
		if strings.TrimSpace(call.Provider) == "" {
			return nil, &FirecrawlError{Message: fmt.Sprintf("exchange call %d: provider is required", i)}
		}
		if strings.TrimSpace(call.Capability) == "" {
			return nil, &FirecrawlError{Message: fmt.Sprintf("exchange call %d: capability is required", i)}
		}
	}
	if opts != nil && opts.Timeout != nil {
		if *opts.Timeout <= 0 {
			return nil, &FirecrawlError{Message: "timeout must be positive"}
		}
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, time.Duration(*opts.Timeout+5000)*time.Millisecond)
		defer cancel()
	}

	body := map[string]interface{}{"exchange": calls}
	mergeOptions(body, opts)

	if _, ok := body["origin"]; !ok {
		body["origin"] = "go-sdk@" + Version
	}
	requestID := ""
	if opts != nil {
		requestID = opts.RequestID
	}
	if requestID == "" {
		var bytes [16]byte
		if _, err := rand.Read(bytes[:]); err != nil {
			return nil, err
		}
		requestID = hex.EncodeToString(bytes[:])
	}
	if !regexp.MustCompile(`^[A-Za-z0-9._:-]{1,128}$`).MatchString(requestID) {
		return nil, &FirecrawlError{Message: "invalid request ID"}
	}
	raw, err := c.http.post(ctx, "/v2/scrape", body, map[string]string{"x-request-id": requestID})
	if err != nil {
		return nil, &ExchangeExecutionError{RequestID: requestID, Err: err}
	}

	var envelope struct {
		ScrapeID string `json:"scrape_id"`
		Data     *struct {
			Exchange    []ExchangeScrapeResult `json:"exchange"`
			CreditsCost *int                   `json:"creditsCost"`
		} `json:"data"`
	}
	if err := json.Unmarshal(raw, &envelope); err != nil {
		return nil, &ExchangeExecutionError{RequestID: requestID, Err: err}
	}
	if envelope.Data == nil || envelope.Data.Exchange == nil || envelope.Data.CreditsCost == nil || *envelope.Data.CreditsCost < 0 {
		return nil, &ExchangeExecutionError{RequestID: requestID, Err: &FirecrawlError{Message: "invalid exchange response"}}
	}
	return &ExchangeScrapeData{
		RequestID:   requestID,
		ScrapeID:    envelope.ScrapeID,
		Exchange:    envelope.Data.Exchange,
		CreditsCost: *envelope.Data.CreditsCost,
	}, nil
}

// FindTools explores the catalogue without executing the tools it returns.
func (c *Client) FindTools(ctx context.Context, opts *FindToolsOptions) (*FindToolsData, error) {
	options := map[string]interface{}{}
	mergeOptions(options, opts)
	result, err := c.ScrapeExchange(ctx, []ExchangeCall{{Provider: "firecrawl-contextual-discovery", Capability: "discovery/context", Options: options}}, nil)
	if err != nil {
		return nil, err
	}
	fail := func(err error) (*FindToolsData, error) {
		return nil, &ExchangeExecutionError{RequestID: result.RequestID, Err: err}
	}
	if len(result.Exchange) != 1 {
		return fail(&FirecrawlError{Message: "missing Find Tools result"})
	}
	item := result.Exchange[0]
	if item.Error != nil {
		return fail(&FirecrawlError{ErrorCode: item.Error.Code, Message: item.Error.Message})
	}
	raw, err := json.Marshal(item.Data)
	if err != nil {
		return fail(err)
	}
	var data FindToolsData
	if err := json.Unmarshal(raw, &data); err != nil {
		return fail(err)
	}
	return &data, nil
}

File: apps/js-sdk/firecrawl/src/v2/methods/tools.ts
import type {
  ExchangeCall,
  ExchangeOptions,
  ExchangeScrapeData,
  ExchangeScrapeResult,
  FindToolsData,
  FindToolsOptions,
} from "../types";
import { SdkError } from "../types";
import { HttpClient } from "../utils/httpClient";
import {
  normalizeAxiosError,
  throwForBadResponse,
} from "../utils/errorHandler";

const EXCHANGE_MAX_CALLS = 10;

function prepareExchangePayload(
  calls: ExchangeCall[],
  opts: ExchangeOptions,
): Record<string, unknown> {
  if (!Array.isArray(calls) || calls.length === 0) {
    throw new Error("exchange requires at least one call");
  }
  if (calls.length > EXCHANGE_MAX_CALLS) {
    throw new Error(`exchange accepts at most ${EXCHANGE_MAX_CALLS} calls`);
  }
  const exchange = calls.map((call, index) => {
    if (!call || typeof call.provider !== "string" || !call.provider.trim()) {
      throw new Error(`exchange[${index}].provider cannot be empty`);
    }
    if (typeof call.capability !== "string" || !call.capability.trim()) {
      throw new Error(`exchange[${index}].capability cannot be empty`);
    }
    if (
      Object.keys(call).some(
        (key) => !["provider", "capability", "options"].includes(key),
      )
    )
      throw new Error("Unknown exchange call option");
    const item: Record<string, unknown> = {
      provider: call.provider.trim(),
      capability: call.capability.trim(),
    };
    if (call.options != null) {
      if (typeof call.options !== "object" || Array.isArray(call.options)) {
        throw new Error(`exchange[${index}].options must be an object`);
      }
      item.options = call.options;
    }
    return item;
  });
  if (
    opts.timeout != null &&
    (!Number.isInteger(opts.timeout) || opts.timeout <= 0)
  ) {
    throw new Error("timeout must be a positive integer");
  }
  const payload: Record<string, unknown> = { exchange };
  if (opts.timeout != null) payload.timeout = opts.timeout;
  if (opts.integration && opts.integration.trim())
    payload.integration = opts.integration.trim();
  if (opts.origin) payload.origin = opts.origin;
  return payload;
}

export async function scrapeExchange(
  http: HttpClient,
  calls: ExchangeCall[],
  opts: ExchangeOptions = {},
): Promise<ExchangeScrapeData> {
  const payload = prepareExchangePayload(calls, opts);
  const requestId = opts.requestId ?? crypto.randomUUID();
  if (!/^[A-Za-z0-9._:-]{1,128}$/.test(requestId))
    throw new Error("Invalid requestId");
  try {
    const res = await http.post<{
      success: boolean;
      scrape_id?: string;
      data?: { exchange?: ExchangeScrapeResult[]; creditsCost?: number };
      error?: string;
    }>("/v2/scrape", payload, {
      headers: { "x-request-id": requestId },
      ...(opts.timeout != null ? { timeoutMs: opts.timeout + 5000 } : {}),
    });
    if (res.status !== 200 || !res.data?.success) {
      throwForBadResponse(res, "exchange");
    }
    const data = res.data.data;
    if (
      !data ||
      !Array.isArray(data.exchange) ||
      typeof data.creditsCost !== "number" ||
      !Number.isInteger(data.creditsCost) ||
      data.creditsCost < 0
    ) {
      throw new SdkError("Invalid exchange response");
    }
    return {
      scrapeId: res.data.scrape_id ?? "",
      requestId,
      exchange: data.exchange,
      creditsCost: data.creditsCost,
    };
  } catch (err: any) {
    try {
      if (err?.isAxiosError) normalizeAxiosError(err, "exchange");
      throw err;
    } catch (error) {
      if (error && typeof error === "object")
        Object.assign(error, { requestId });
      throw error;
    }
  }
}

export async function findTools(
  http: HttpClient,
  options: FindToolsOptions = {},
): Promise<FindToolsData> {
  const result = await scrapeExchange(http, [
    {
      provider: "firecrawl-contextual-discovery",
      capability: "discovery/context",
      options: { ...options },
    },
  ]);
  const item = result.exchange[0];
  if (!item) throw new SdkError("Missing Find Tools result");
  if (item.error)
    throw Object.assign(
      new SdkError(item.error.message, item.error.status, item.error.code),
      { requestId: result.requestId },
    );
  return item.data as FindToolsData;
}
